content/blog/2019-12-20-Internal-Function-Memoization/index.md

```diff
@@ -0,0 +1,175 @@
+- Base Cases
+- Loop induction variable and upper bound
+- How these recursive calls are combined at return (i.e., fibonacci adds the results from the f(n-1) and f(n-2) together)
+- The offset from our induction variable in the recursive calls (i.e., (n-1), (n-3), etc.)
+
+Information collection posed many challenges due to the low-level of LLVMIR, and desire to generalize our solution.
+#### Finding the Base Cases
+To more formally define the term _base cases_, we required base cases to be comparisons with the input argument to the function, found at the top of a program, that resulted in returning a value within the body guarded by the comparison.
+While humans have no trouble deciphering the base case given a program, it is much more challenging to find them when sifting through LLVMIR. We used an ad-hoc approach where we found the constant return values and then walked up the CFG to find the conditional statements that would have led to those constant outputs. From the condition instruction we extracted the function argument that would have led to that constant to be returned, and we end up with a list of argument-return pairs, which sufficiently describes the base cases of the function.
+
+Doing this better would involve abstract interpretation. For a particular function argument, it could trigger a recursive call or it could not. With abstract interpretation, we could find all values of the argument such that the basic blocks with recursive function calls are not traversed. Those would be the base cases. 
+
+
+#### Finding the Loop Induction Variable
+For this project, we simply initialized the induction variable at the largest base case argument and incremented until it reached the function call's argument.
+
+#### Finding the instructions that combine the recursive calls
+We keep a running list of all recursive call instructions. For all instructions in the recursive function that use the result of a recursive call as an operand, we clone them and add the clone to a list of _dependents_ as they depend on the recursive calls. We clone instructions here so we can delete all instructions in the recursive function and insert our own. 
+
+#### Finding Recursive Call Offsets
+For each recursive call that we find, we look at the call argument and assert that it is a constant offset from the original function argument---i.e., of the form $(n-c)$ where $n$ is the original function argument and $c$ is a constant. We remember that $c$. 
+
+### Creating a Model for Iterative Programs to use Collected Information
+Once we have the information from the recursive program, we need to find a model such that given this information, can generate a iterative program. To determine what output we wanted, we wrote a iterative fibonacci program. Instead of writing a single transformation function $f$(Input Recursive Program) = Output Iterative Program, we break down the programs according to a layout. 
```

Comment:
```suggestion
Once we have the information from the recursive program, we need to find a model such that given this information, can generate an iterative program. To determine what output we wanted, we wrote an iterative fibonacci program. Instead of writing a single transformation function $f$(Input Recursive Program) = Output Iterative Program, we break down the programs according to a layout. 
```